### PR TITLE
[Core] Fix log buffering issue

### LINKED ISSUE #2065: [Core] Logging buffered for run section
When running the following yaml file, the `hi2` and `hi3` will only be printed after the `sleep` is finished. It can cause trouble for spot jobs debugging as the preemption can cause the latest log not showing up.
```
run: |
  echo hi
  echo hi2
  echo hi3
  sleep 12000
  echo bye
```

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #2065.

The original selectors implementation introduces buffer. We changed it to threading based implementation and it now fixes the buffering problem. Tried asyncio, but since the log_lib can be used in a thread and asyncio does not work well within a thread, we now use the threading implementation.

TODO:
- [x] Examine if the previous arguments for `process_subprocess_stream` is still useful

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch --cloud gcp --cpus 2+ -c test-log test.yaml` with test.yaml in the issue.
  - [x] `sky launch -c test-log-2 --num-nodes 2 --cloud gcp --cpus 2+ test.yaml`
  - [x]  Progress bar test in #261 `sky launch -c progress-bar test2.yaml`
```
# test.yaml
run: |
  echo hi
  echo hi2
  echo hi3
  sleep 12000
  echo bye
```
```
# test2.yaml
name: minimal

setup: |
  echo "running setup"

run: |
  conda env list
  python3 -u -c "from tqdm import tqdm; import time; [time.sleep(1) for i in tqdm(range(10))]"
  echo bye
```
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/skylet/log_lib.py [hunk 1]
```diff
@@ -26,84 +27,98 @@
...
+        # different thread.
+        # selectors is possible to handle the multiplexing of stdout/stderr,
+        # but it introduces buffering making the output not streaming.
+        with multiprocessing.pool.ThreadPool(processes=1) as pool:
+            err_args = copy.copy(args)
+            err_args.line_processor = None
+            stderr_fut = pool.apply_async(_handle_io_stream,
+                                          args=(proc.stderr, sys.stderr,
+                                                err_args))
+            stdout = _handle_io_stream(proc.stdout, sys.stdout, args)
```
Comment (romilbhardwaj) [inline sky/skylet/log_lib.py:117]: Nit: Any particular reason to not also launch stdout processing as a separate thread and then call `pool.join()` on both futures? Just curious..

File: sky/skylet/log_lib.py [hunk 2]
```diff
@@ -26,84 +27,98 @@
...
+            stderr = stderr_fut.get()
```
Comment (romilbhardwaj) [inline sky/skylet/log_lib.py:118]: Do you think we also need to do any exception handling here (or after `while True:` at L70) if something goes wrong in _handle_io_stream? Or will this run normally even if there's an exception?

**Review by romilbhardwaj — APPROVED** [review]
Thanks for catching and fixing this critical bug @Michaelvll! LGTM. I tried looking into solutions to somehow fix buffering for selectors and setting `os.O_NONBLOCK` on the stderr and stdout descriptors, but it didn't help. This solution seems to be a good fix.

Tested with the YAML from #2065 and explicit output to stderr with `>&2 echo "I am stderr"`.

File: sky/skylet/log_lib.py (hunk 1 above)
Comment (Michaelvll) [reply sky/skylet/log_lib.py:120]: We cannot launch a new thread for the stdout, as the rich.status does not work in a thread other than the main thread, which is used in `log_utils.RayUpLineProcessor`. Just added a comment here for future reference. Thanks!

File: sky/skylet/log_lib.py (hunk 2 above)
Comment (Michaelvll) [reply sky/skylet/log_lib.py:121]: I could not think of any possible error that can be raised there. Maybe we can just let the error to be thrown as we did before, and wait for any report of exceptions?

**Comment by Michaelvll** [comment]
Merged the #1982. Testing again:
- [ ] `pytest tests/test_smoke.py`

**Pull request merged**